juju/model.py

```diff
@@ -0,0 +1,210 @@
+        # since they may have different views of the units of that app. However, they should be considered
+        # equivalent, since they are views of the same entity. Generally, this should only matter for using
+        # model.app as an index into a given relation's relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Application):
+            return False
+        return self.name == other.name
+
+
+class Unit:
+    def __init__(self, name):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            return False
+        return self.name == other.name
+
+
+class RelationMap(Mapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, local_app, local_unit, backend=None):
+        self._relation_names = relation_names
+        self._local_app = local_app
+        self._local_unit = local_unit
+        self._backend = backend or ModelBackend()
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __iter__(self):
+        return iter(self._relation_names)
+
+    def __len__(self):
+        return len(self._relation_names)
+
+    @lru_cache(maxsize=None)
+    def __getitem__(self, relation_name):
+        # Lazy-load the list of relations for this relation name.
+        relation_ids = self._backend.relation_ids(relation_name)
+        return [Relation(relation_name, relation_id, self._local_app, self._local_unit, self._backend) for relation_id in relation_ids]
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, local_app, local_unit, backend=None):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend or ModelBackend()
+        self.data = RelationData(self.relation_name, relation_id, local_app, local_unit, self._backend)
+
+    @property
+    @lru_cache(maxsize=None)
+    def apps(self):
+        # Lazy-load the apps on this relation.
+        unit_names = self._backend.relation_list(self.relation_id)
+        apps = {}
+        for unit_name in unit_names:
+            app_name = unit_name.split('/')[0]
+            app = apps.get(app_name)
+            if not app:
+                app = apps[app_name] = Application(app_name, [])
+            app.units.append(Unit(unit_name))
```

Comment:
It seems cleaner to create the list of units upfront, and then create the Application with the correct list of units instead of an empty one. Otherwise, any logic that we put inside the Application to process that list of units anyhow will be wrong.

That said, per earlier note in the review, instead of "injecting" into the Application its data from the outside, it might be even cleaner to hand the relation_id to the Application itself, via a proper API (something like add_relation, noting an Application may be present in multiple relations), and let the Application figure its own life out.

Later we'll also need to de-dup the Application, so that in the model every individual application is present in the model only once. We can do that in a follow up so this PR with the main chunk can go in sooner rather than later.

Reply:
> That said, per earlier note in the review, instead of "injecting" into the Application its data from the outside, it might be even cleaner to hand the relation_id to the Application itself, via a proper API (something like add_relation, noting an Application may be present in multiple relations), and let the Application figure its own life out.

+1 for moving in this direction

> Later we'll also need to de-dup the Application, so that in the model every individual application is present in the model only once. We can do that in a follow up so this PR with the main chunk can go in sooner rather than later.

There will still be cases where the same "application" is represented more than once in the model as views via different relations.  However, the `(name, relation)` pair can certainly be de-duped.  Do you have a strong opinion on an explicit factory method (`Application.get()`) vs using `Application.__new__` to do the de-duping?  I like the slightly cleaner syntax of the latter, since it's the same as constructing the instance normally, but it does hide the fact that you might end up with an existing instance rather than a new one.  I've seen both patterns widely used in Python, however.

Reply:
I also think that if I'm going to be refactoring the loading of `Units` into `Application` anyway by assuming some manner of relation ID, it would make sense to address the de-dupe logic at the same time.